csrc/host_ir/jit.cpp

```diff
@@ -681,74 +733,124 @@ class HostIrCompileDispatcher : public OptInDispatch {
-    builder_.CreateCall(set_tensor_func, {out_tensor, in_tensor});
-
-    // Bind the output tensor to val_to_value
-    val_to_value_[out_tv] = out_tensor;
+    llvm::LLVMContext& context = builder_.getContext();
+  
+    if (out_tv->hasRoot()) {
+      std::optional<std::vector<int64_t>> permutation =
+          ir_utils::computePermutation(
+              out_tv->getRootDomain(), out_tv->getLogicalDomain());
+      NVF_ERROR(
+          permutation.has_value(),
+          "The logical domain of a Set.Permute is supposed to be a permutation"
+          " of the root domain: ",
+          out_tv);
+
+      llvm::Function* permute_out_func =
+          module->getFunction(kPermuteOutFuncName);
+      
+      // Create array of permutation values
+      llvm::ArrayType* perm_array_type = getInt64StaticArrayType(context, permutation.value().size());
+      llvm::Value* perm_array = builder_.CreateAlloca(perm_array_type, nullptr, "permutation");
+
+      for (size_t i = 0; i < permutation.value().size(); ++i) {
+        llvm::Value* gep = builder_.CreateInBoundsGEP(
+            perm_array_type, perm_array, 
+            {builder_.getInt32(0), builder_.getInt32(i)});
+        builder_.CreateStore(builder_.getInt64(permutation.value()[i]), gep);
+      }
+      
+      llvm::Type* int64_ptr_type = getInt64PtrType(context);
+      llvm::Value* perm_ptr = builder_.CreateBitCast(perm_array, int64_ptr_type);
+      llvm::Value* perm_size = builder_.getInt64(permutation.value().size());
+      if(out_tensor != nullptr) {
+        builder_.CreateCall(permute_out_func, {out_tensor, in_tensor, perm_ptr, perm_size});
+        return;
+      }
+             out_tensor = builder_.CreateCall(
+           module->getFunction(kPermuteFuncName), {in_tensor, perm_ptr, perm_size}, "permute");
+       val_to_value_[static_cast<Val*>(out_tv)] = out_tensor;
+       return;
+    }
+    if(out_tensor != nullptr) {
+      llvm::Function* set_tensor_out_func =
+          module->getFunction(kSetTensorOutFuncName);
+      builder_.CreateCall(set_tensor_out_func, {out_tensor, in_tensor});
+      return;
+    }
+    out_tensor = builder_.CreateCall(
+        module->getFunction(kSetTensorFuncName), {in_tensor}, "set");
+    val_to_value_[static_cast<Val*>(out_tv)] = out_tensor;
   }
 
+
   void handle(MatmulOp* matmul_op) final {
     llvm::Module* module = builder_.GetInsertBlock()->getParent()->getParent();
 
-    llvm::Value* t_a =
-        getOrCreateValue(matmul_op->inA(), val_to_value_, builder_);
-    llvm::Value* t_b =
-        getOrCreateValue(matmul_op->inB(), val_to_value_, builder_);
-    llvm::Value* t_out =
-        getOrCreateValue(matmul_op->out(), val_to_value_, builder_);
-
-    builder_.CreateCall(
-        module->getFunction(kMatmulOutFuncName), {t_out, t_a, t_b});
-    val_to_value_[matmul_op->out()] = t_out;
+    llvm::Value* a =
+        getOrDefault(val_to_value_, static_cast<Val*>(matmul_op->inA()));
```

Comment:
```suggestion
        getOrDefault(val_to_value_, matmul_op->inA());
```
upcast is done automatically -- no need for a static_cast